File: Data Structures & Algorithms/subsets/submission-0.cpp
class Solution {
public:

    void solve(int index, vector<int>& tmp, vector<vector<int>> &ans, vector<int>& nums){
        if(index == nums.size()){
            ans.push_back(tmp);
            return;
        }
        tmp.push_back(nums[index]);
        solve(index+1, tmp, ans, nums);
        tmp.pop_back();
        solve(index+1, tmp, ans, nums);
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> tmp;
        solve(0, tmp, ans, nums);
        return ans;
    }
};

File: Data Structures & Algorithms/subsets-ii/submission-0.java
class Solution {

    private void solve(int index, int[] nums, HashSet<List<Integer>> set, List<Integer> tmp){
        if(index == nums.length){
            set.add(new ArrayList<>(tmp));
            return;
        }

        for(int i=index; i<nums.length; i++){
            tmp.add(nums[i]);
            solve(i+1, nums, set, tmp);
            tmp.remove(tmp.size()-1);
            solve(i+1, nums, set, tmp);
        }
    }

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<Integer> tmp = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        HashSet<List<Integer>> set = new HashSet<>();

        solve(0, nums, set, tmp);

        for(List<Integer> it: set)
            ans.add(it);

        return ans;
    }
}
